ide/che-core-ide-jseditor/src/main/java/org/eclipse/che/ide/jseditor/client/debug/BreakpointManagerImpl.java

```diff
@@ -309,50 +370,270 @@ public void onLineChange(final VirtualFile file, final int firstLine, final int
-                    LOG.fine("Moving breakpoint " + breakpoint + " delta=" + delta);
-                    toRemove.add(breakpoint);
-                    toAdd.add(new Breakpoint(breakpoint.getType(), breakpoint.getLineNumber() + delta,
-                                             breakpoint.getPath(), breakpoint.getFile(), breakpoint.getMessage()));
+
+                toRemove.add(breakpoint);
+                toAdd.add(new Breakpoint(breakpoint.getType(),
+                                         breakpoint.getLineNumber() + delta,
+                                         breakpoint.getPath(),
+                                         breakpoint.getFile(),
+                                         breakpoint.getMessage(),
+                                         breakpoint.isActive()));
+            }
+
+            Debugger debugger = debuggerManager.getDebugger();
+            if (debugger != null) {
+                for (final Breakpoint breakpoint : toRemove) {
+                    deleteBreakpoint(file, debugger, breakpoint);
+                }
+                for (final Breakpoint breakpoint : toAdd) {
+                    if (isCodeExecutable(file, breakpoint.getLineNumber())) {
+                        addBreakpoint(file, debugger, breakpoint.getLineNumber());
+                    }
                 }
             }
-            for (final Breakpoint breakpoint : toRemove) {
-                changeBreakPointState(breakpoint.getLineNumber());
+        }
+    }
+
+    /**
+     * Registers events handlers.
+     */
+
+    private void registerEventHandlers(EventBus eventBus) {
+        eventBus.addHandler(EditorOpenedEvent.TYPE, new EditorOpenedEventHandler() {
+            @Override
+            public void onEditorOpened(EditorOpenedEvent event) {
+                onOpenEditor(event.getFile(), event.getEditor());
+            }
+        });
+
+        eventBus.addHandler(DebuggerStateEvent.TYPE, new DebuggerStateEventHandler() {
+            @Override
+            public void onStateChanged(DebuggerStateEvent event) {
+                BreakpointManagerImpl.this.debuggerState = event.getState();
+                if (event.isConnectedState()) {
+                    onDebuggerConnected(event.getDebugger());
+                } else if (event.isDisconnectedState()) {
+                    onDebuggerDisconnected();
+                } else if (event.isInitializedState()) {
+                    onDebugInitialized(event.getDebugger());
+                }
             }
-            for (final Breakpoint breakpoint : toAdd) {
-                changeBreakPointState(breakpoint.getLineNumber());
+        });
+    }
+
+    /**
+     * Debugger has been initialized.
+     * Method adds stored breakpoints.
+     */
+    private void onDebugInitialized(Debugger debugger) {
+        Storage localStorage = Storage.getLocalStorageIfSupported();
+        if (localStorage != null) {
```

Comment:
It would be better to check if local storage is null then simple call return instead of creating nested levels